File: Java/DoublyCircularLinkedList.java
package Java;


// compile with: javac Java/DoublyCircularLinkedList.java 
// then, run with: java Java/DoublyCircularLinkedList
class DoublyNode {
    public Object data;

    public DoublyNode prev;
    public DoublyNode next;


    public DoublyNode(Object data) {
        this.data = data;
    }
}


public class DoublyCircularLinkedList {
    private DoublyNode head;
    private DoublyNode tail;
    
    private int size;


    public DoublyNode getHead() {
        return this.head;
    }


    public DoublyNode getTail() {
        return this.tail;
    }


    public int getSize() {
        return this.size;
    }


    public void addElementToStart(Object data) {
        DoublyNode newHeadNode = new DoublyNode(data);

        if (this.size == 0) {
            this.head = newHeadNode;
            this.tail = newHeadNode;

            newHeadNode.next = newHeadNode;
            newHeadNode.prev = newHeadNode;
        }
        else {
            this.head.prev = newHeadNode;
            newHeadNode.next = this.head;
            newHeadNode.prev = this.tail;
            this.tail.next = newHeadNode;
            this.head = newHeadNode;
        }
        
        this.size++;
    }


    public void deleteFirstElement() {
        if (this.size == 0) {
            return;
        }
        
        DoublyNode newHeadNode = this.head.next;
        newHeadNode.prev = this.tail;
        this.tail.next = newHeadNode;
        this.head = newHeadNode;

        this.size--;
    }


    public void addElementToEnd(Object data) {
        DoublyNode newTailNode = new DoublyNode(data);

        if (this.size == 0) {
            this.head = newTailNode;
            this.tail = newTailNode;

            newTailNode.next = newTailNode;
            newTailNode.prev = newTailNode;
        }
        else {
            this.tail.next = newTailNode;
            newTailNode.prev = this.tail;
            newTailNode.next = this.head;
            this.head.prev = newTailNode;
            this.tail = newTailNode;
        }

        this.size++;
    }


    public void deleteLastElement() {
        if (this.size == 0) {
            return;
        }
        
        DoublyNode newTailNode = this.tail.prev;
        newTailNode.next = this.head;
        this.head.prev = newTailNode;
        this.tail = newTailNode;

        this.size--;
    }


    public static void main(String[] args) {}
}

File: TypeScript/DoublyCircularLinkedList.ts
// to run this file, use this command: npx ts-node TypeScript/DoublyCircularLinkedList.ts
class DoublyNode {
    data!: any;

    prev!: DoublyNode | null;
    next!: DoublyNode | null;


    constructor(data: any) {
        this.data = data;
    }
}


class DoublyCircularLinkedList {
    head!: DoublyNode | null;
    tail!: DoublyNode | null;
    
    size!:number;


    getHead(): DoublyNode | null {
        return this.head;
    }


    getTail(): DoublyNode | null {
        return this.tail;
    }


    getSize(): number {
        return this.size;
    }


    addElementToStart(data: any): void {
       const new_head_node = new DoublyNode(data);

        if (this.size == 0) {
            this.head = new_head_node;
            this.tail = new_head_node;

            new_head_node.next = new_head_node;
            new_head_node.prev = new_head_node;
        }
        else {
            this.head!.prev = new_head_node;
            new_head_node.next = this.head;
            new_head_node.prev = this.tail;
            this.tail!.next = new_head_node;
            this.head = new_head_node;
        }

        this.size++;
    }


    deleteFirstElement(): void {
        if (this.size == 0) {
            return;
        }
        
        const new_head_node = this.head!.next;
        new_head_node!.prev = this.tail;
        this.tail!.next = new_head_node;
        this.head = new_head_node;

        this.size--;
    }


    addElementToEnd(data: any): void {
        const new_tail_node = new DoublyNode(data);

        if (this.size == 0){
            this.head = new_tail_node;
            this.tail = new_tail_node;

            new_tail_node.next = new_tail_node;
            new_tail_node.prev = new_tail_node;
        }
        else {
            this.tail!.next = new_tail_node;
            new_tail_node.prev = this.tail;
            new_tail_node.next = this.head;
            this.head!.prev = new_tail_node;
            this.tail = new_tail_node;
        }

        this.size++;
    }


    deleteLastElement(): void {
        if (this.size == 0) {
            return;
        }
        
        const new_tail_node = this.tail!.prev;
        new_tail_node!.next = this.head;
        this.head!.prev = new_tail_node;
        this.tail = new_tail_node;

        this.size--;
    }
}